File: scalechain-transaction/src/test/scala/io/scalechain/blockchain/transaction/SignatureTestTrait.scala
package io.scalechain.blockchain.transaction

import io.scalechain.blockchain.proto.{Block, Transaction, LockingScript, GenerationTransactionInput, NormalTransactionInput}
import io.scalechain.blockchain.script.ScriptEnvironment
import org.scalatest._

/**
 * Created by kangmo on 11/16/15.
 */
trait SignatureTestTrait extends ShouldMatchers {
  this: Suite =>

  protected def verifyTransactionInput(subject : String, spendingTransaction : Transaction, inputIndex : Int, lockingScript : LockingScript): Unit =
  {
    assert(inputIndex >= 0)
    assert(inputIndex < spendingTransaction.inputs.length)
    val env = new ScriptEnvironment(spendingTransaction, Some(inputIndex))

    val txInput = spendingTransaction.inputs(inputIndex)
    txInput match {
      case normalTxInput : NormalTransactionInput => {
        new NormalTransactionVerifier(normalTxInput).verify(env, lockingScript)
      }
      case generationTxInput : GenerationTransactionInput  => {
        new GenerationTransactionVerifier(generationTxInput).verify(null, null)
      }
    }
  }
}

File: scalechain-transaction/src/test/kotlin/io/scalechain/blockchain/transaction/SignatureTestTrait.kt
package io.scalechain.blockchain.transaction

import io.kotlintest.KTestJUnitRunner
import io.kotlintest.matchers.Matchers
import io.scalechain.blockchain.proto.Block
import io.scalechain.blockchain.proto.Transaction
import io.scalechain.blockchain.proto.LockingScript
import io.scalechain.blockchain.proto.GenerationTransactionInput
import io.scalechain.blockchain.proto.NormalTransactionInput
import io.scalechain.blockchain.script.ScriptEnvironment
import org.junit.runner.RunWith

/**
 * Created by kangmo on 11/16/15.
 */
@RunWith(KTestJUnitRunner::class)
interface SignatureTestTrait : Matchers {

  fun verifyTransactionInput(subject : String, spendingTransaction : Transaction, inputIndex : Int, lockingScript : LockingScript): Unit
  {
    assert(inputIndex >= 0)
    assert(inputIndex < spendingTransaction.inputs.size)
    val env = ScriptEnvironment(spendingTransaction, inputIndex)

    val txInput = spendingTransaction.inputs[inputIndex]
    when {
      txInput is NormalTransactionInput -> {
        // We don't need to access db to get the locking script, as we already have the locking script.
        NormalTransactionVerifier(null/*db*/, txInput, spendingTransaction, inputIndex).verify(env, lockingScript)

      }
      txInput is GenerationTransactionInput -> {
        // The db parameter is not used yet.
        GenerationTransactionVerifier(null/*db*/, txInput).verify(null, null)
      }
    }
  }
}
